sike-java/src/main/java/com/wultra/security/pqc/sike/math/optimized/fp/FpElementOpti.java

```diff
@@ -419,8 +420,9 @@ public boolean equals(Object o) {
         if (!(o instanceof FpElementOpti)) {
             return false;
         }
-        FpElementOpti other = (FpElementOpti) o;
-        return getX().equals(other.getX());
+        FpElementOpti that = (FpElementOpti) o;
+        // Use constant time comparison to avoid timing attacks
+        return org.bouncycastle.util.Arrays.constantTimeAreEqual(getX().toByteArray(), that.getX().toByteArray());
```

Comment:
This is weird - `getX` uses `getEncoded` internally, so we convert back and forth between bytes and `BigInteger`...

Reply:
The tests will fail in case we compare the internal long array representation because multiple long array values can represent the same BigInteger value. This is caused by the Montgomery reduction. I will look into this issue.

Reply:
The public key comparison in optimized C implementation as well as Go implementation compare the standard representation of public keys (the `byte[]` representation), not the Montgomery representation (the optimized `long[]` internal representation). So the comparison is done after conversion from Montgomery representation to the standard representation using the `from_fp2mont()` method, which is the equivalent of the `getEncoded()` method in Sike for Java.

I suggest we stick to the comparison in the standard representation. It is safer to compare exact numeric values than comparing elements in the Montgomery representation which is mainly used for optimization. We would also have to divert the implementation from original logic if we wanted to compare in the Montgomery representation which I think is not a good idea.

Reply:
OK, fair and makes sense.